File: src/main/scala/org/ranceworks/postclustering/token/EnglishTokenizer.scala
package org.ranceworks.postclustering.token

import org.apache.lucene.analysis.TokenStream
import org.apache.lucene.analysis.en.EnglishAnalyzer
import org.apache.lucene.analysis.tokenattributes.CharTermAttribute

import scala.annotation.tailrec

class EnglishTokenizer extends Tokenizer {
  override def tokenize(text: String): List[String] = {
    val analyzer = new EnglishAnalyzer()
    val tokenStream: TokenStream =  analyzer.tokenStream(this.getClass.getName, text)

    val charTermAttribute =  tokenStream.addAttribute(classOf[CharTermAttribute])

    @tailrec
    def findTokens(foundTokens: List[String], ts: TokenStream): List[String] = {
       ts.incrementToken() match {
        case true => findTokens(foundTokens :+ charTermAttribute.toString, ts)
        case false => foundTokens
      }
    }

    tokenStream.reset()
    val tokens = findTokens(List[String](), tokenStream)
    tokenStream.end()
    tokens
  }
}

File: src/main/java/org/ranceworks/postclustering/tokenizer/EnglishTokenizer.java
package org.ranceworks.postclustering.tokenizer;

import com.sun.javafx.scene.control.skin.VirtualFlow;
import org.apache.lucene.analysis.TokenStream;
import org.apache.lucene.analysis.en.EnglishAnalyzer;
import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;


public class EnglishTokenizer  implements  Tokenizer {

    @Override
    public List<String> tokenize(String text) throws IOException {
        final EnglishAnalyzer analyzer = new EnglishAnalyzer();
        final TokenStream ts = analyzer.tokenStream(this.getClass().getName(), text);
        final CharTermAttribute attr =  ts.addAttribute(CharTermAttribute.class);

        final List<String> tokens =  new ArrayList<>();
        ts.reset();
        while(ts.incrementToken()) {
            tokens.add(attr.toString());
        }
        ts.end();
        return tokens;
    }
}
